jobs/populate_db.py

```diff
@@ -44,48 +56,40 @@ def insert_into_db(data_set):
             4b. remove_fields will check to see if an incomplete task exists, and will remove it if it does
         At this point the process is complete and the next tool is passed to check_for_entry.
     """
+    logger.info("Begin tool processing and db insertion.")
     for tool in data_set:
-        check_for_entry(tool)
-    return "All tools have been processed."
+        add_or_update_tool(tool)
+    logger.info("Tool processing complete.")
+    return
 
 
-def check_for_entry(tool):
-    """Receives a tool and checks to see if an entry exists in the DB."""
-    tool_name = tool["name"]
-    result = db.session.execute(select(Tool).where(Tool.name == tool_name)).all()
-    # If an entry exists move on to the deprecation check
-    if len(result) > 0:
-        print(f"{tool_name} already exists in database.")
-    else:
-        add_tool_entry(tool)
-    check_deprecation(tool)
-
-
-def add_tool_entry(tool):
+def add_or_update_tool(tool):
```

Comment:
add_or_update_tool should do only what it says, i.e. take a tool and update the tools db table. Whether a tool should be inserted or not (because it's deprecated, for instance) should be checked before calling this function. This may involve rethinking how/when deprecated tools and tasks are addressed in the pipeline. There is also at least one edge case that is not being handled right now: currently, a deprecated tool is inserted into the db upon first encounter.

More generally, functions should either return something, or produce a side effect. In this case, add_or_update_tool produces a side effect: updating the db. This is where its responsibility should end. Calling a function from inside another function is fine – as long as the called function returns something, but a pipeline should be made up of a sequence of atomic functions, i.e. each function in the sequence should be called after the previous function has completed.

Reply:
I dealt with the chaining issue and turned insert_into_db into a master function, so to speak.  Each function should now (I think -- sometimes I worry that I'm a little too close to the code to see it entirely clearly) be properly atomic.   [6667818](https://github.com/wikimedia/toolhunt/pull/51/commits/6667818d9de929b733791e4ab9b3366aca578054)